test/functional/lua/with_spec.lua

```diff
@@ -0,0 +1,74 @@
+local t = require('test.testutil')
+local n = require('test.functional.testnvim')()
+
+local exec_lua = n.exec_lua
+
+before_each(function()
+  n.clear()
+end)
+
+describe('vim._with resets', function()
+  it('buffer', function()
+    local ids = exec_lua [[
+      prev = vim.api.nvim_get_current_buf()
+      vim.cmd.new()
+      before = vim.api.nvim_get_current_buf()
+      vim._with({buf = prev}, function() end)
+      after = vim.api.nvim_get_current_buf()
+      return {before, after}
+    ]]
+    t.eq(ids[1], ids[2])
+  end)
+
+  it('window', function()
+    local ids = exec_lua [[
+      prev = vim.api.nvim_get_current_win()
+      vim.cmd.new()
+      before = vim.api.nvim_get_current_win()
+      vim._with({win = prev}, function() end)
+      after = vim.api.nvim_get_current_win()
+      return {before, after}
+    ]]
+    t.eq(ids[1], ids[2])
+  end)
```

Comment:
These two tests only check the `vim._with()` restores original window/buffer. They don't check if the current window/buffer is indeed the given ones when the code is executed.